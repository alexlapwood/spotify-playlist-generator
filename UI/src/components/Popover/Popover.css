.Popover {
  background-color: var(--accent-lighten);
  border-radius: 3px;
  box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.2), 0 3px 6px 0 rgba(0, 0, 0, 0.2);
  color: var(--accent-darken-2);
  max-width: 300px;
  position: relative;
  padding: 1.5rem;
  z-index: 500;
}

.Popover--fullwidth {
  border-radius: 0;
}

.Popover-arrow--bottom,
.Popover-arrow--top {
  height: 8px;
  width: calc(8px * 2);
}

.Popover-arrow--left,
.Popover-arrow--right {
  height: calc(8px * 2);
  width: 8px;
}

.Popover-arrow--bottom:before,
.Popover-arrow--bottom:after,
.Popover-arrow--left:before,
.Popover-arrow--left:after,
.Popover-arrow--right:before,
.Popover-arrow--right:after,
.Popover-arrow--top:before,
.Popover-arrow--top:after {
  content: "";
  position: absolute;
  z-index: 500;
}

.Popover-arrow--bottom:before {
  border-bottom: 8px solid rgba(0, 0, 0, 0.2);
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
}

.Popover-arrow--bottom:after {
  border-bottom: 8px solid var(--accent-lighten);
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  transform: translateY(2px);
}

.Popover-arrow--left:before {
  border-bottom: 8px solid transparent;
  border-left: 8px solid rgba(0, 0, 0, 0.2);
  border-top: 8px solid transparent;
}

.Popover-arrow--left:after {
  border-bottom: 8px solid transparent;
  border-left: 8px solid var(--accent-lighten);
  border-top: 8px solid transparent;
  transform: translateX(-2px);
}

.Popover-arrow--right:before {
  border-bottom: 8px solid transparent;
  border-right: 8px solid rgba(0, 0, 0, 0.2);
  border-top: 8px solid transparent;
}

.Popover-arrow--right:after {
  border-bottom: 8px solid transparent;
  border-right: 8px solid var(--accent-lighten);
  border-top: 8px solid transparent;
  transform: translateX(2px);
}

.Popover-arrow--top:before {
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-top: 8px solid rgba(0, 0, 0, 0.2);
}

.Popover-arrow--top:after {
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-top: 8px solid var(--accent-lighten);
  transform: translateY(-2px);
}
