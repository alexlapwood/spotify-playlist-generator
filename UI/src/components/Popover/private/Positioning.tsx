import React from "react";
import { Portal } from "react-portal";

import PositioningHelper, { Location } from "./helpers/positioning";

/**
 * Checks equality by iterating through keys on an object and returning false when any key has
 * values which are not strictly equal.
 *
 * @param object1 The first object to compare.
 * @param object2 The second object to compare.
 *
 * @returns {boolean} True when the objects pass a shallow compare, false when the objects fail a
 * shallow compare.
 */
function shallowCompare(
  object1?: { [index: string]: any },
  object2?: { [index: string]: any }
) {
  return (
    object1 &&
    object2 &&
    Object.keys(object1).length === Object.keys(object2).length &&
    Object.keys(object1).every((key) => object1[key] === object2[key])
  );
}

interface Props {
  children: (location?: Location, isFullWidth?: boolean) => React.ReactNode;
  pageBuffer?: number;
  preferredLocation?: Location;
  triggerBuffer?: number;
  triggerRef: React.RefObject<HTMLElement>;
}

interface State {
  isFullWidth?: boolean;
  location?: Location;
  style?: React.CSSProperties;
}

export default class Positioning extends React.Component<Props, State> {
  private ref = React.createRef<HTMLDivElement>();
  private updatePositionTimeoutId?: ReturnType<typeof setTimeout>;
  private windowHasResizeListener = false;

  state: State = {};

  public componentDidMount() {
    this.updatePosition();

    window.addEventListener("resize", this.updatePosition);
  }

  public componentWillUnmount() {
    window.removeEventListener("resize", this.updatePosition);
  }

  public componentDidUpdate() {
    this.updatePosition();
  }

  public render() {
    const { children } = this.props;
    const { isFullWidth, location, style } = this.state;
    return (
      <Portal>
        <div ref={this.ref} style={{ position: "absolute", ...style }}>
          {children(location, isFullWidth)}
        </div>
      </Portal>
    );
  }

  /* eslint-disable no-param-reassign */
  private getContentRect = (contentElement: HTMLElement) => {
    const previousStyle = {
      left: contentElement.style.left,
      top: contentElement.style.top,
    };
    /**
     * We place the contentElement at the top left of the page before retrieving its measurements so
     * that the size is not influenced by the edges of the page.
     */
    contentElement.style.left = "0";
    contentElement.style.top = "0";

    const contentRect = contentElement.getBoundingClientRect();

    contentElement.style.left = previousStyle.left;
    contentElement.style.top = previousStyle.top;

    return contentRect;
  };
  /* eslint-enable no-param-reassign */

  private updatePosition = () => {
    const { pageBuffer, preferredLocation, triggerBuffer } = this.props;

    /**
     * Running `updatePosition` regularly accommodates the following situations:
     * - The content resizes (could be handled with a ResizeObserver)
     * - The trigger resizes (could be handled with a ResizeObserver)
     * - The trigger changes position (currently we have no way of observing this reliably)
     */
    if (this.updatePositionTimeoutId) {
      clearTimeout(this.updatePositionTimeoutId);
    }
    setTimeout(this.updatePosition, 100);

    if (!this.ref.current || !this.props.triggerRef.current) {
      return;
    }

    const contentRect = this.getContentRect(this.ref.current);
    const triggerRect = this.props.triggerRef.current.getBoundingClientRect();

    const positionHelper = new PositioningHelper(
      preferredLocation,
      contentRect,
      triggerRect,
      triggerBuffer,
      pageBuffer
    );

    const location = positionHelper.getLocation();
    const style = positionHelper.getPositionStyle();
    const isFullWidth = positionHelper.isFullWidth();

    // This function is called often so we only call `setState` if something has changed.
    if (
      location !== this.state.location ||
      !shallowCompare(style, this.state.style)
    ) {
      this.setState({
        isFullWidth,
        location,
        style,
      });
    }
  };
}
