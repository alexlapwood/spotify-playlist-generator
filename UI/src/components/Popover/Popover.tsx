import cn from "classnames";
import React from "react";

import getFocusableElement from "./private/helpers/getFocusableElement";
import { Location } from "./private/helpers/positioning";
import PortalFocusHelper from "./private/PortalFocusHelper";
import Positioning from "./private/Positioning";

import "./Popover.css";

interface Props {
  children?: React.ReactNode;
  className?: string;
  id?: string;
  onClickOutside?: () => void;
  preferredPosition?: Location;
  triggerRef: React.RefObject<
    | HTMLElement
    | {
        rootNode: HTMLElement | React.RefObject<HTMLElement>;
      }
  >;
}

export default class Popover extends React.Component<Props> {
  private wrapperRef = React.createRef<HTMLDivElement>();

  public componentDidMount() {
    this.addEventListeners();
  }

  public componentDidUpdate() {
    this.removeEventListeners();
    this.addEventListeners();
  }

  public componentWillUnmount() {
    this.removeEventListeners();

    if (
      this.wrapperRef.current &&
      this.wrapperRef.current.contains(document.activeElement)
    ) {
      this.focusOnTheTrigger();
    }
  }

  public render() {
    const {
      className,
      children,
      id,
      onClickOutside,
      preferredPosition,
    } = this.props;

    const triggerRef = this.getTriggerRef();

    const arrowSize = 8;
    const borderShadowSize = 1;

    return (
      <React.Fragment>
        <Positioning
          pageBuffer={arrowSize}
          preferredLocation={preferredPosition}
          triggerBuffer={arrowSize + borderShadowSize}
          triggerRef={triggerRef}
        >
          {(location?: Location, isFullWidth?: boolean) => (
            <React.Fragment>
              <div
                className={cn(
                  `Popover`,
                  isFullWidth && "Popover--fullwidth",
                  className
                )}
                id={id}
                ref={this.wrapperRef}
                role="dialog"
              >
                <PortalFocusHelper
                  focusPortalRef={triggerRef}
                  onReturnFocus={onClickOutside}
                >
                  <div className={cn("Popover-body")}>{children}</div>
                </PortalFocusHelper>
              </div>
              <Positioning
                pageBuffer={arrowSize}
                preferredLocation={location}
                triggerRef={triggerRef}
              >
                {() => <div className={`Popover-arrow--${location}`} />}
              </Positioning>
            </React.Fragment>
          )}
        </Positioning>
      </React.Fragment>
    );
  }

  private addEventListeners = () => {
    document.addEventListener("click", this.handleClickOutside);
  };

  private removeEventListeners = () => {
    document.removeEventListener("click", this.handleClickOutside);
  };

  private handleClickOutside = (
    event: MouseEvent | FocusEvent | React.FocusEvent
  ) => {
    const { onClickOutside } = this.props;

    if (
      onClickOutside &&
      event.target &&
      this.wrapperRef.current &&
      !this.wrapperRef.current.contains(event.target as Node)
    ) {
      onClickOutside && onClickOutside();
    }
  };

  private focusOnTheTrigger = () => {
    const triggerRef = this.getTriggerRef();

    if (!triggerRef.current) {
      return;
    }

    const focusableElement = getFocusableElement(triggerRef.current);

    if (!focusableElement) {
      return;
    }

    focusableElement.focus();
  };

  getTriggerRef = () => {
    return {
      _triggerRef: this.props.triggerRef,
      get current() {
        if (!this._triggerRef.current) {
          return null;
        }

        if (this._triggerRef.current instanceof HTMLElement) {
          return this._triggerRef.current;
        }
        if (!this._triggerRef.current.rootNode) {
          return null;
        }

        if (this._triggerRef.current.rootNode instanceof HTMLElement) {
          return this._triggerRef.current.rootNode;
        }

        if (this._triggerRef.current.rootNode.current instanceof HTMLElement) {
          return this._triggerRef.current.rootNode.current;
        }

        return null;
      },
    };
  };
}
