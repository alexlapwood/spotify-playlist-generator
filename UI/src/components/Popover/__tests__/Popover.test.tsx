import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";

import Popover from "../Popover";
import Positioning from "../private/Positioning";

jest.mock("uuid/v4", () => jest.fn(() => "123"));

describe("<Popover />", () => {
  it("renders without crashing", () => {
    // Arrange
    const triggerRef = React.createRef<HTMLElement>();
    const wrapper = mount(
      <Popover id="test-popover" triggerRef={triggerRef} />
    );

    // Assert
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("calls onClickOutside when a click event occurs outside the popover", () => {
    // Arrange
    const onClickOutsideMock = jest.fn();
    const triggerRef = React.createRef<HTMLElement>();
    mount(
      <Popover
        id="test-popover"
        onClickOutside={onClickOutsideMock}
        triggerRef={triggerRef}
      />
    );

    // Act
    const clickEvent = new MouseEvent("click");
    document.dispatchEvent(clickEvent);

    // Assert
    expect(onClickOutsideMock).toHaveBeenCalled();
  });

  it("closing the popover from within the popover returns the focus to the trigger", () => {
    // Arrange
    const triggerFocusMock = jest.fn();
    const closeButtonRef = React.createRef<HTMLButtonElement>();
    const trigger = document.createElement("button");
    const triggerRef = { current: trigger };
    triggerRef.current.addEventListener("focus", triggerFocusMock);
    const wrapper = mount(
      <Popover id="test-popover" triggerRef={triggerRef}>
        <button ref={closeButtonRef} type="button" />
      </Popover>
    );

    // Act
    closeButtonRef.current?.focus();
    wrapper.instance().componentWillUnmount?.();

    // Assert
    expect(triggerFocusMock).toHaveBeenCalled();
  });

  it("closing the popover from outside the popover does not return the focus to the trigger", () => {
    // Arrange
    const triggerFocusMock = jest.fn();
    const trigger = document.createElement("button");
    const triggerRef = { current: trigger };
    triggerRef.current.addEventListener("focus", triggerFocusMock);
    const wrapper = mount(
      <Popover id="test-popover" triggerRef={triggerRef} />
    );

    // Act
    wrapper.instance().componentWillUnmount?.();

    // Assert
    expect(triggerFocusMock).not.toHaveBeenCalled();
  });

  describe("triggerRef", () => {
    it("proxies the rootNode callback ref of class components", () => {
      // Arrange
      const trigger = document.createElement("button");
      const triggerClassComponent = { rootNode: trigger };
      const triggerRef = { current: triggerClassComponent };
      const wrapper = mount(
        <Popover id="test-popover" triggerRef={triggerRef} />
      );

      // Assert
      expect(wrapper.find(Positioning).first().props().triggerRef.current).toBe(
        trigger
      );
    });

    it("proxies the rootNode ref object of class components", () => {
      // Arrange
      const trigger = document.createElement("button");
      const triggerClassComponent = { rootNode: { current: trigger } };
      const triggerRef = { current: triggerClassComponent };
      const wrapper = mount(
        <Popover id="test-popover" triggerRef={triggerRef} />
      );

      // Assert
      expect(wrapper.find(Positioning).first().props().triggerRef.current).toBe(
        trigger
      );
    });
  });
});
