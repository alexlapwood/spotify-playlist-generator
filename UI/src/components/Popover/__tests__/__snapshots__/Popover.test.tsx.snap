// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popover /> renders without crashing 1`] = `
<Popover
  id="test-popover"
  triggerRef={
    Object {
      "current": null,
    }
  }
>
  <Positioning
    pageBuffer={8}
    triggerBuffer={9}
    triggerRef={
      Object {
        "_triggerRef": Object {
          "current": null,
        },
        "current": null,
      }
    }
  >
    <Portal>
      <Portal
        containerInfo={
          <div>
            <div
              style="position: absolute;"
            >
              <div
                class="Popover"
                id="test-popover"
                role="dialog"
              >
                <div
                  tabindex="0"
                />
                <div
                  tabindex="0"
                />
                <div>
                  <div
                    class="Popover-body"
                  />
                </div>
                <div
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        }
      >
        <div
          style={
            Object {
              "position": "absolute",
            }
          }
        >
          <div
            className="Popover"
            id="test-popover"
            role="dialog"
          >
            <PortalFocusHelper
              focusPortalRef={
                Object {
                  "_triggerRef": Object {
                    "current": null,
                  },
                  "current": null,
                }
              }
            >
              <div
                onFocus={[Function]}
                tabIndex={0}
              />
              <div
                onFocus={[Function]}
                tabIndex={0}
              />
              <div>
                <div
                  className="Popover-body"
                />
              </div>
              <div
                onFocus={[Function]}
                tabIndex={0}
              />
            </PortalFocusHelper>
          </div>
          <Positioning
            triggerRef={
              Object {
                "_triggerRef": Object {
                  "current": null,
                },
                "current": null,
              }
            }
          >
            <Portal>
              <Portal
                containerInfo={
                  <div>
                    <div
                      style="position: absolute;"
                    >
                      <div
                        class="Popover-arrow--undefined"
                      />
                    </div>
                  </div>
                }
              >
                <div
                  style={
                    Object {
                      "position": "absolute",
                    }
                  }
                >
                  <div
                    className="Popover-arrow--undefined"
                  />
                </div>
              </Portal>
            </Portal>
          </Positioning>
        </div>
      </Portal>
    </Portal>
  </Positioning>
</Popover>
`;
